kimchi/src/circuits/polynomials/foreign_field_add/witness.rs

```diff
@@ -0,0 +1,346 @@
+/// all necessary values needed for the final bound check witness. Meaning:
+/// - the bound as a ForeignElement
+/// - the carry_lo and carry_mi values
+fn compute_bound_values<F: PrimeField>(
+    result: &ForeignElement<F, 3>,
+    modulus: &ForeignElement<F, 3>,
+) -> (ForeignElement<F, 3>, F, F) {
+    let max = BigUint::from(TWO_TO_LIMB).pow(3);
+    let big_mod = modulus.to_big();
+    let bound = result.to_big() + max - big_mod;
+    let bound_limbs = ForeignElement::<F, 3>::from_biguint(bound);
+    let carry_mi = *bound_limbs.hi() - *result.hi() + *modulus.hi() - F::from(TWO_TO_LIMB);
+    let carry_lo =
+        *bound_limbs.mi() - *result.mi() + *modulus.mi() + carry_mi * F::from(TWO_TO_LIMB);
+    (bound_limbs, carry_lo, carry_mi)
+}
+*/
+
+/// Given a left and right inputs to an addition or subtraction, and a modulus, it computes
+/// all necessary values needed for the witness layout. Meaning:
+/// - the result of the addition/subtraction as a ForeignElement
+/// - the sign of the operation
+/// - the overflow flag
+/// - the carry_lo and carry_mi values
+fn compute_subadd_values<F: PrimeField>(
+    left_input: &ForeignElement<F, 3>,
+    right_input: &ForeignElement<F, 4>,
+    add: bool,
+    foreign_modulus: &ForeignElement<F, 3>,
+) -> (ForeignElement<F, 3>, F, F, F, F) {
+    let two_to_limb = F::from(TWO_TO_LIMB);
+
+    // Compute bigint version of the inputs
+    let left = left_input.to_big();
+    let right = right_input.to_big();
+    let right_hi = right_input.limbs[3] * two_to_limb + right_input.limbs[2]; // This allows to store 2^88 in the high limb
+
+    let modulus = foreign_modulus.to_big();
+
+    if add {
+        // addition
+        let sig = F::one();
+        let sum = left + right;
+        let overflows = sum >= modulus;
+        let ovf = if overflows { F::one() } else { F::zero() };
+        let res = if overflows { sum - modulus } else { sum };
+        let res_limbs = ForeignElement::from_biguint(res);
+        let carry_mi = *res_limbs.hi() - *left_input.hi() - right_hi + ovf * *foreign_modulus.hi();
+        let carry_lo = *res_limbs.mi() - *left_input.mi() - right_input.limbs[1]
+            + ovf * foreign_modulus.mi()
+            + two_to_limb * carry_mi;
+        (res_limbs, sig, ovf, carry_lo, carry_mi)
+    } else {
+        // subtraction
+        let sig = -F::one();
+        let underflows = left < right;
+        let udf = if underflows { -F::one() } else { F::zero() };
+        let res = if underflows {
+            modulus + left - right
+        } else {
+            left - right
+        };
+        let res_limbs = ForeignElement::from_biguint(res);
+        let carry_mi = *res_limbs.hi() - *left_input.hi() + right_hi + udf * *foreign_modulus.hi();
+        let carry_lo = *res_limbs.mi() - *left_input.mi()
+            + right_input.limbs[1]
+            + udf * foreign_modulus.mi()
+            + two_to_limb * carry_mi;
+        (res_limbs, sig, udf, carry_lo, carry_mi)
+    }
```

Comment:
Yeah, something like this... but somehow the other approach also looks quite clean

Reply:
actually, if you want to simplify things, instead of this
`    let field_overflow = if !has_overflow {
        // None
        F::zero()
    } else {
        if add {
            // Overflow
            F::one()
        } else {
            // Underflow
            -F::one()
        }
    };`
    
 you can have this 😉 

`    let field_overflow = if has_overflow { sign } else { F::zero() };`

Reply:
Unluckily we cannot reduce the result to just one line because of types incompatibility, otherwise we could have just `    result = left + sign * right - field_overflow * modulus`